File: DataStructures/bit manipulations/leetcode/89. Gray Code/GPT/Gray_Code.py
# ## 1. 問題分析結果

# ### 競技プログラミング視点

# * 生成長は **M = 2^n**。隣接（先頭⇔末尾含む）で **1 bit** だけ異なる巡回列（Gray code）。
# * 最速は定番式 **`G(i) = i ^ (i >> 1)`** を `i=0..M-1` に適用するだけ。

#   * **時間**: O(M)
#   * **追加メモリ**: O(1)（出力配列は不可避）
# * Python では **リスト内包表記**が最速（C 実装の `range` 走査＋list確保）。

# ### 業務開発視点

# * 可読・保守：`grayCode(n)` の **純関数**で完結。命名とドキュメンテーションで意図を明確化。
# * 型安全：`n: int` と戻り値 `List[int]` を明示。
# * エラーハンドリング：LeetCode は制約内入力が前提だが、実務では整数・範囲チェックを追加可能（コメントで例示）。

# ### Python特有考慮

# * **CPython** の `list(range(M))`/内包表記は高速。
# * **GIL** の影響は無関係（単一スレッド・CPU 計算）。
# * **内蔵データ構造**：可変長 `list` を一度だけ割り当てて終了。中間オブジェクトを作らない。

# ---

# ## 2. アルゴリズム比較表

# | アプローチ                         | 時間計算量       | 空間計算量                   | Python実装コスト | 可読性 | 標準ライブラリ活用 | CPython最適化 | 備考           |
# | ----------------------------- | ----------- | ----------------------- | ----------- | --- | --------- | ---------- | ------------ |
# | 方法A：`i ^ (i >> 1)` 直接生成（内包表記） | **O(2^n)**  | **O(2^n)**（出力）/ O(1) 追加 | 低           | 高   | `range`   | 適          | 最速・最小追加メモリ   |
# | 方法B：反射（鏡映）法                   | O(2^n)      | O(2^n)                  | 中           | 中   | なし        | 適          | 段階反転＋上位ビット付与 |
# | 方法C：DFS バックトラッキング             | O(2^n·n) 以上 | O(2^n)                  | 低           | 低   | なし        | 不適         | 実利に乏しい（遅い）   |

# ---

# ## 3. 実装方針

# * **選択**：方法A（`G(i) = i ^ (i >> 1)`）
# * **理由**：計算量最良、コード最短、Python では **リスト内包が最速**。
# * **最適化**：`range(M)` を一度走査し、内包表記で結果を **一括構築**。中間配列や関数呼び出しオーバーヘッドを避ける。

# ---

# ## 4. 実装コード（LeetCode 提出用 / Class 形式）

# ```python
from __future__ import annotations

from typing import List


class Solution:
    """
    Gray Code generator (n-bit).
    LeetCode 提出想定のクラス形式実装（純関数・副作用なし）。
    """

    def grayCode(self, n: int) -> List[int]:
        """
        Return any valid n-bit Gray code sequence of length 2^n.
        Uses the standard formula: G(i) = i ^ (i >> 1)

        Args:
            n (int): number of bits (1 <= n <= 16)

        Returns:
            List[int]: Gray code sequence starting with 0, length = 2^n

        Complexity:
            Time: O(2^n)
            Space: O(2^n)  # output list
        """
        # 実務では以下のような防御的検証を加えてもよい（LeetCode 制約下では不要）
        # if not isinstance(n, int):
        #     raise TypeError("n must be int")
        # if n < 1 or n > 16:
        #     raise ValueError("n must be in [1, 16]")

        m: int = 1 << n
        # リスト内包表記は CPython で高速。中間オブジェクトを作らない。
        result: List[int] = [(i ^ (i >> 1)) for i in range(m)]
        return result


# ```

# ---

# ## 5. 検証（方針のみ・コード不要）

# * **境界値**：`n=1` → `[0,1]`、`n=16` → 長さ 65536 の整合性（先頭/末尾の 1bit 差）。
# * **隣接 1bit 差チェック**：`popcount(result[i] ^ result[i+1]) == 1`（末尾→先頭含む）。
# * **型チェック**：`pylance` で `List[int]` 戻り・引数 `int` の一致を確認。

File: DataStructures/bit manipulations/leetcode/89. Gray Code/GPT/Gray_Code.js
// # 1. 問題の分析

// ## 競技プログラミング視点での分析

// * 生成すべき配列長は **M = 2^n**。各連続要素（最後と最初を含む）が **1bit** だけ異なる巡回列（Gray code）。
// * 代表解は **i ⊕ (i >> 1)** により `i=0..M-1` を一発生成（辞書式 Gray code）。
//   これで隣接ペアは常に 1bit 差となり、先頭 0 と末尾も 1bit 差を満たす。

// ## 業務開発視点での分析

// * **保守性/可読性**：`grayCode(n)` という単機能・副作用なしの純関数に集約。
// * **例外方針**：`TypeError`（非整数/非数）、`RangeError`（制約 1 ≤ n ≤ 16 を外れる）を早期に投げる。
// * **命名**：`M` を総要素数、`res` を結果配列として固定。コメントは最小限かつ要点のみ。

// ## JavaScript特有の考慮点

// * **V8最適化**：

//   * 配列を **事前確保（new Array(M))** → pushの形状変化を避け hidden class を安定化。
//   * 単純な `for` ループ（クロージャ/高階関数の割当コストを回避）。
// * **GC対策**：一時オブジェクト生成なし。数値のみの単型配列を維持。
// * **bit演算**：`n ≤ 16` なので `1 << n` は安全（32bit 範囲内）。

// ---

// # 2. アルゴリズムアプローチ比較

// | アプローチ                      | 時間計算量     | 空間計算量    | JS実装コスト | 可読性 | 備考                              |
// | -------------------------- | --------- | -------- | ------- | --- | ------------------------------- |
// | 方法A：ビット式 `i ^ (i>>1)` 直接生成 | **O(M)**  | **O(M)** | 低       | 中   | 最速・最小メモリ（必要配列のみ）。隣接1bit差を自明に満たす |
// | 方法B：反転鏡映法（反射して上位bit付与）     | O(M)      | O(M)     | 中       | 高   | 各段で既列を反転コピーし上位bitセット            |
// | 方法C：バックトラック（1bit差制約でDFS）   | O(M·n) 以上 | O(M)     | 低       | 低   | 実装は簡単だが最悪遅い、枝刈りが必要              |

// ※ M = 2^n

// ---

// # 3. 選択したアルゴリズムと理由

// * **選択**：方法A（ビット式 `i ^ (i>>1)`）
// * **理由**：

//   * **計算量**：単一ループ **O(M)** で最短。
//   * **メモリ**：結果配列のみ **O(M)**。
//   * **実装効率**：一行で Gray 値が得られ、V8 に優しい単純 for＋事前確保。
// * **JS特有の最適化ポイント**：

//   * `const M = 1 << n; const res = new Array(M);` で固定長・単型を維持。
//   * ホットパス外で入力検証を完了し、ループは純粋算術のみ。

// ---

// # 4. コード実装（solution.js）

// ```js
"use strict";

/**
 * Generate n-bit Gray code sequence.
 * Pure function: no side effects.
 *
 * @param {number} n - Number of bits (integer in [1, 16]).
 * @returns {number[]} Gray code sequence of length 2^n.
 * @throws {TypeError}  If n is not a finite integer.
 * @throws {RangeError} If n is out of allowed range [1, 16].
 *
 * @description
 * Constructs Gray codes via the standard formula: G(i) = i ^ (i >> 1),
 * for i = 0..(2^n - 1).
 *
 * @complexity Time O(2^n), Space O(2^n)
 */
function grayCode(n) {
  // --- Input validation (fast fail, outside of hot path) ---
  if (typeof n !== "number" || !Number.isFinite(n) || (n | 0) !== n) {
    throw new TypeError("n must be a finite integer");
  }
  if (n < 1 || n > 16) {
    throw new RangeError("n must be in [1, 16]");
  }

  // --- Precompute size and preallocate result array ---
  const M = 1 << n; // safe for n <= 16 (<= 65536)
  const res = new Array(M);

  // --- Single pass: Gray value via i ^ (i >> 1) ---
  for (let i = 0; i < M; i++) {
    res[i] = (i ^ (i >>> 1)) >>> 0; // >>>0 to keep unsigned 32-bit (not required, but stable)
  }

  return res;
}

module.exports = { grayCode };
// ```

// ---

// # 5. 追加メモ（JS最適化チェックリスト）

// * ループは **for** を使用（`map/forEach` は避ける）。
// * 配列は **事前確保**＋**単型（number のみ）** を維持。
// * 一時オブジェクトやクロージャを作らない。
// * 例外系はホットパス外で早期チェック。
// * bit演算は 32bit 整数に収まる範囲で実施（本件は `n ≤ 16` で安全）。
